File: IT/2/Multithreading/Esercizi/Java/Esercizio9.java
/* file Esercizio9.java */

import java.util.Arrays;
import java.util.List;
import java.util.ArrayList;
import java.util.concurrent.Executors;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.Future;
import java.util.concurrent.Callable;
import java.util.concurrent.TimeUnit;

class Target implements Callable<int[]> {
	private int index;

	public Target(int index) {
		this.index = index;
	}

	@Override
	public int [] call() throws Exception {
		int [] row = new int[10];

		for (int i = 0; i < 10; i++) {
			row[i] = index;
		}

		for (int index = 0, k = 1; index < 10; index++, k++) {
			row[index] *= k;
		}

		return row;
	}
}

public class Esercizio9 {
	public static void main(String [] args) {
		int nThreads = 10;

		ExecutorService threadPool = Executors.newFixedThreadPool(nThreads);
		List<Future<int[]>> tbp = new ArrayList<>();

		for (int i = 1; i <= nThreads; i++) {
			Future<int[]> row = threadPool.submit(new Target(i));
			tbp.add(row);
		}

		threadPool.shutdown();
		try {
			threadPool.awaitTermination(1, TimeUnit.SECONDS);
		}
		catch (InterruptedException e) {
			e.printStackTrace();
		}

		try {
			for (Future<int[]> row : tbp) {
				System.out.println(Arrays.toString(row.get()));
			}
		}
		catch (ExecutionException e) {
			e.printStackTrace();
		}
		catch (InterruptedException e) {
			e.printStackTrace();
		} 
	}
}

File: IT/2/Multithreading/Esercizi/C++/esercizio9.cpp
/*
file esercizio9.cpp
compilare con: g++ -std=c++11 esercizio9.cpp -lpthread
*/

#include <iostream>
#include <list>
#include <iterator>
#include <future>

using std::cin;
using std::cout;
using std::endl;
using std::copy;
using std::list;
using std::ostream_iterator;
using std::async;
using std::future;

int* func(int index) {
	int* row = new int[10];

	for (int i = 0; i < 10; i++) {
		row[i] = index;
	}

	for (int index = 0, k = 1;  index < 10; index++, k++) {
		row[index] *= k;
	}

	return row;
}

int main()
{
	int nThreads = 10;

	list<future<int*>> tbp;

	for (int i = 1; i <= nThreads; i++) {
		tbp.push_back(async(std::launch::async, func, i));
	}

	for (auto it = tbp.begin(); it != tbp.end(); ++it) {
		int *row = (*it).get();
		copy(row, row+10, ostream_iterator<int>(cout, " "));
		cout << endl;
		delete row;
	}
}

